app/helpers/coins_helper.rb

```diff
@@ -49,4 +50,76 @@ def coins_serializer(coins)
       methods: %i[sparkline price market_cap change1h change24h change7d volume24h]
     )
   end
+
+  def seconds_to_next_day
+    1.day.since.beginning_of_day - Time.now
+  end
+
+  def seconds_to_next_hour
+    1.hour.since.beginning_of_hour - Time.now
+  end
+
+  def latest_total_market_cap
+    distribute_reads(max_lag: MAX_ACCEPTABLE_REPLICATION_LAG, lag_failover: true) do
+      latest_market_metric = MarketMetric.latest
+      latest_market_metric.total_market_cap if latest_market_metric.present?
+    end
+  end
+
+  def historical_total_market_data(days: 7, months: nil)
+    distribute_reads(max_lag: MAX_ACCEPTABLE_REPLICATION_LAG, lag_failover: true) do
+      if days.present?
+        MarketMetric.daily(days)
+      elsif months.present?
+        MarketMetric.monthly(months)
+      end
+    end
+  end
+
+  def live_market_dominance(number_of_other_coins: 4, no_cache: false)
+    total_market_cap = latest_total_market_cap
+
+    distribute_reads(max_lag: MAX_ACCEPTABLE_REPLICATION_LAG, lag_failover: true) do
+      pinned_coin_slug = 'bitcoin'
+      top_coins = Coin.listed.legit.top(number_of_other_coins).where.not(slug: pinned_coin_slug)
+      pinned_coin = Coin.where(slug: pinned_coin_slug)
+      coins = Coin.from("((#{top_coins.to_sql}) UNION (#{pinned_coin.to_sql})) AS coins")
+      market_dominance = coins.map do |coin|
+        market_cap = coin.market_cap || 0
+        market_percentage = total_market_cap.present? ? market_cap / total_market_cap : 0
+        [coin.coin_key, {
+          :id => coin.id,
+          :name => coin.name,
+          :symbol => coin.symbol,
+          :slug => coin.slug,
+          :price_usd => coin.price || 0,
+          :market_percentage => market_percentage
+        }]
+      end.to_h
+
+      Rails.cache.write("market_dominance/#{number_of_other_coins}", market_dominance, expires_in: seconds_to_next_day) unless no_cache || total_market_cap.blank?
+
+      market_dominance
+    end
+  end
+
+  def market_dominance(number_of_other_coins: 4)
```

Comment:
Hm, it's normally best practice to combine the `live_market_dominance` method with the `market_dominance` method via `Rails.cache.fetch` and just pass in the `force` parameter if we want to refresh the cache, i.e.

```
def market_dominance(number_of_other_coins: 4, no_cache: false)
  total_market_cap = latest_total_market_cap

  distribute_reads(max_lag: MAX_ACCEPTABLE_REPLICATION_LAG, lag_failover: true) do
    pinned_coin_slug = 'bitcoin'
    top_coins = Coin.listed.legit.top(number_of_other_coins).where.not(slug: pinned_coin_slug)
    pinned_coin = Coin.where(slug: pinned_coin_slug)
    coins = Coin.from("((#{top_coins.to_sql}) UNION (#{pinned_coin.to_sql})) AS coins")
    market_dominance = coins.map do |coin|
      market_cap = coin.market_cap || 0
      market_percentage = total_market_cap.present? ? market_cap / total_market_cap : 0
      [coin.coin_key, {
        :id => coin.id,
        :name => coin.name,
        :symbol => coin.symbol,
        :slug => coin.slug,
        :price_usd => coin.price || 0,
        :market_percentage => market_percentage
      }]
    end.to_h

    Rails.cache.fetch("market_dominance/#{number_of_other_coins}", expires_in: seconds_to_next_day, force: no_cache) do
      market_dominance
    end
  end
end
```